### PR TITLE
vim: Add support for `ap` and `ip` paragraph text objects

### PR DESCRIPTION
**Pull request by noritada**
This PR adds support for `ap`/`ip` text objects in Vim mode and allows users to perform paragraph-based operations.

Cases where compatibility with Neovim's behavior is checked, cases where there are known differences in behavior with Neovim (cases where the landing position is other than the beginning of the line), and cases where the Neovim behavior in the test suite seems strange are separated in the test code so that they can be identified.

Release Notes:

- Added support for `ap` and `ip` paragraph text objects in Vim mode ([#7359](https://github.com/zed-industries/zed/issues/7359)).

### CODE REVIEW

**Comment by ConradIrwin** [comment]
@noritada thank you for taking a pass at this!

I have a number of suggestions for improvement. If you'd like to work together on getting this over the line, feel free to book time here: https://calendly.com/conradirwin/pairing; otherwise I'll upload comments to Github and we can try to do this async.

**Comment by noritada** [comment]
Thank you very much! I've booked at Calendly.

File: crates/vim/src/normal/delete.rs [hunk 1]
```diff
@@ -96,6 +96,23 @@ pub fn delete_object(vim: &mut Vim, object: Object, around: bool, cx: &mut Windo
...
+                                (offset_range.end + '\n'.len_utf8()).to_display_point(map);
+                        }
+
+                        let end_at_eof = map.chars_at(selection.end).next().is_none();
+                        if end_at_eof && selection.start.row() > 0 {
+                            should_move_to_start.insert(selection.id);
+                            selection.start =
+                                (offset_range.start - '\n'.len_utf8()).to_display_point(map);
+                        }
+                    }
```
Comment (ConradIrwin) [inline crates/vim/src/normal/delete.rs:115]: It looks like this code does the same as the previous if condition. Can we make the condition more specific and share the code?

File: crates/vim/src/object.rs [hunk 2]
```diff
@@ -812,6 +954,190 @@ mod test {
...
+            cx.assert_binding_matches_all(keystrokes, TRAILING_NORMAL_PARAGRAPH_EXAMPLE)
+                .await;
+        }
+        // Nvim in the test suite seems to be returning strange landing position.
+        cx.assert_binding_matches_all_exempted(
+            ["d", "i", "p"],
+            TRAILING_NORMAL_PARAGRAPH_EXAMPLE,
+            ExemptionFeatures::IncorrectLandingPosition,
+        )
+        .await;
```
Comment (ConradIrwin) [inline crates/vim/src/object.rs:1062]: It looks like the tests are correctly emulating neovim here? It seems like maybe the difference arises depending on having a trailing newline on the buffer:

```
\n
ˇParagraph\n
```

In that case `d i p` will give you:

```
\n
ˇ\n
```

However if you have no trailing new line 

```
\n
ˇParagraph
```

vim gives you:

```
ˇ\n
```

**Comment by noritada** [comment]
@ConradIrwin I have updated the main part. Could you review it?

"Paragraphs" in vim include not only non-blank paragraphs but also "blank paragraphs," which is a little different from what most people think of as paragraphs. I felt that it would be difficult to understand the code without shared recognition of this specification, so I have documented the specification in the doc comment.

I thought it would not be good to tinker with the delete edge case without confirmation of the main part, so I have pushed only the main part first.

File: crates/vim/src/object.rs [hunk 3]
```diff
@@ -449,6 +456,141 @@ fn expand_to_include_whitespace(
...
+    Break(DisplayPoint),
+    EndOfFile(DisplayPoint),
+}
+
+fn paragraph(
+    map: &DisplaySnapshot,
+    relative_to: DisplayPoint,
+    around: bool,
+) -> Option<Range<DisplayPoint>> {
+    let mut backward_iter = map.reverse_chars_at(relative_to);
```
[the remark this answers is not in the document]
Comment (noritada) [reply crates/vim/src/object.rs]: I think changing from the block snapshot to the buffer snapshot has been completed for the main part (but works for normal/delete.rs is still necessary).

**Comment by ConradIrwin** [comment]
@noritada Thanks for this, I like the new approach!

A few notes:
* In practice it seems like `vim_mode` style paragraphs need completely different code from the existing paragraph code in the editor. I think it would be better to make our own copies of those functions `vim_paragraph_end` and put them in the vim crate instead of adding the boolean. (I had hoped they would be more similar, but I was wrong).
* I think the `target_visual_mode` should return Mode::VisualLine for these (`When used in Visual mode it is made linewise.` from the docs). That would explain the test failures for visual mode.

**Comment by ConradIrwin** [comment]
@noritada Please let me know if you're still planning to work on this.

**Comment by noritada** [comment]
@ConradIrwin 
I think I have completed all of the necessary changes to the code that you pointed out.

However, after investigating the issue of not being able to properly test `v i p` and `v a p`, I found that it was due to the way we test visual line mode, so I am attempting to fix it. Since this fix involves a large change on its own, I have split the changes to PR #8333.

Could you please take a look at this PR to see why tests for the visual line mode is not passing?

**Comment by ConradIrwin** [comment]
@noritada I think even with the fixes to the tests, there are still bugs to be addressed here:

```
## vip should work on the last line of a file with no trailing newline:

# initial state:
ˇThe quick brown fox jumps over the lazy dog.
# keystrokes:
v i p
# neovim state:
«Tˇ»he quick brown fox jumps over the lazy dog.
# zed state:
ˇThe quick brown fox jumps over the lazy dog.
```

```
## vip should select the last line-ending inside the current paragraph

# initial state:
ˇThe quick brown fox jumps
 over the lazy dog.

# keystrokes:
v i p
# neovim state:
«The quick brown fox jumps
 ˇ»over the lazy dog.

# zed state:
«The quick brown fox jumps
ˇ» over the lazy dog.
```

In this case the zed line-mode selection does not extend to the last line of the paragraph.

It is probably easier to create test-cases manually instead of using `assert_binding_matches_all`. That is a little overkill (though it's kind of nice for fuzz-testing, it's quite hard to develop against as you can't keep track of progress). I would suggest making some more concrete tests like:

```
#[gpui::test]
fn test_visual_paragrah_object(cx: &mut gpui:::TestAppContext) {
        cx.assert_shared_state(indoc! {
            "ˇThe quick brown
            fox jumps over
            the lazy dog
            "})
        .await;
        cx.simulate_shared_keystrokes(["v", "i", "p"]).await;

        cx.assert_shared_state(indoc! {
            "«ˇThe quick brown
            fox jumps over
            t»he lazy dog
            "})
        .await;
}
```

Happy to work together on this if you want help getting it over the line: https://calendly.com/conradirwin/pairing.

**Comment by noritada** [comment]
Thank you very much! I understand the policy.

By building and using the app, I have understood the problem and have fixed the end position of the selection.
I will also make changes for edge cases and make the code testable.

File: crates/vim/src/object.rs [hunk 4]
```diff
@@ -607,6 +614,99 @@ fn expand_to_include_whitespace(
...
+fn paragraph(
+    map: &DisplaySnapshot,
+    relative_to: DisplayPoint,
+    around: bool,
+) -> Option<Range<DisplayPoint>> {
+    let mut paragraph_start = start_of_paragraph(map, relative_to);
+    let mut paragraph_end = end_of_paragraph(map, relative_to);
+
+    let paragraph_end_row = paragraph_end.row();
+    let paragraph_ends_with_eof = paragraph_end_row == map.max_buffer_row();
```
Comment (ConradIrwin) [inline crates/vim/src/object.rs:641]: @noritada max_buffer_row is in the wrong co-ordinate system (c.f. #8081).

```suggestion
    let paragraph_ends_with_eof = paragraph_end_row == map.max_point().row();
```

**Comment by ConradIrwin** [comment]
@noritada One comment for you, but this seems to work well for me manually. Is there anything else you want to do before I merge?

**Comment by noritada** [comment]
@ConradIrwin Sorry for my absence for a few days.

There was an issue with `v i p` for only one blank "paragraph" resulting in the selection including a  subsequent line and I wanted to fix it and I have fixed this issue.
Thank you for your patience. I am glad you are merging.

File: crates/vim/src/object.rs (hunk 4 above)
Comment (noritada) [reply crates/vim/src/object.rs]: Thank you very much! I have incorporated this change.

**Comment by ConradIrwin** [comment]
Thanks for this @noritada!

**Pull request merged**